src/ossl/mldsa.rs

```diff
@@ -0,0 +1,533 @@
+                        | CKH_HEDGE_REQUIRED
+                        | CKH_DETERMINISTIC_REQUIRED => (),
+                        _ => return Err(CKR_MECHANISM_PARAM_INVALID)?,
+                    }
+                    mldsa_params.hedge = params.hedgeVariant;
+                    if params.ulContextLen > 0 {
+                        mldsa_params.context = Some(bytes_to_vec!(
+                            params.pContext,
+                            params.ulContextLen
+                        ));
+                    }
+                    mldsa_params.hash = params.hash;
+                }
+                _ => return Err(CKR_MECHANISM_INVALID)?,
+            }
+        }
+        Ok(mldsa_params)
+    }
+
+    fn ossl_params(&self) -> Result<OsslParam> {
+        let mut params = OsslParam::with_capacity(1);
+        if let Some(ctx) = &self.context {
+            params.add_octet_string(
+                name_as_char(OSSL_SIGNATURE_PARAM_CONTEXT_STRING),
+                ctx,
+            )?;
+        }
+        if self.hedge == CKH_DETERMINISTIC_REQUIRED {
+            params.add_owned_int(
+                name_as_char(OSSL_SIGNATURE_PARAM_DETERMINISTIC),
+                1,
+            )?;
+        }
+        params.finalize();
+
+        Ok(params)
+    }
+}
+
+#[derive(Debug)]
+pub struct MlDsaOperation {
+    mech: CK_MECHANISM_TYPE,
+    finalized: bool,
+    in_use: bool,
+    params: MlDsaParams,
+    sigctx: EvpPkeyCtx,
+    data: Option<Vec<u8>>,
+}
+
+impl MechOperation for MlDsaOperation {
+    fn mechanism(&self) -> Result<CK_MECHANISM_TYPE> {
+        Ok(self.mech)
+    }
+
+    fn finalized(&self) -> bool {
+        self.finalized
+    }
+
+    #[cfg(feature = "fips")]
+    fn fips_approved(&self) -> Option<bool> {
+        Some(true)
+    }
+}
+
+impl MlDsaOperation {
+    pub fn sign_new(
+        mech: &CK_MECHANISM,
+        key: &Object,
+    ) -> Result<MlDsaOperation> {
+        let param_set = key.get_attr_as_ulong(CKA_PARAMETER_SET)?;
+        let mldsa_params = MlDsaParams::new(mech, param_set)?;
+        let mut privkey = EvpPkey::privkey_from_object(key)?;
+        let sigctx = privkey.new_ctx()?;
+        Ok(MlDsaOperation {
+            mech: mech.mechanism,
+            finalized: false,
+            in_use: false,
+            params: mldsa_params,
+            sigctx: sigctx,
+            data: None,
+        })
+    }
+
+    pub fn verify_new(
+        mech: &CK_MECHANISM,
+        key: &Object,
+    ) -> Result<MlDsaOperation> {
+        let param_set = key.get_attr_as_ulong(CKA_PARAMETER_SET)?;
+        let mldsa_params = MlDsaParams::new(mech, param_set)?;
+        let mut pubkey = EvpPkey::pubkey_from_object(key)?;
+        let sigctx = pubkey.new_ctx()?;
+        Ok(MlDsaOperation {
+            mech: mech.mechanism,
+            finalized: false,
+            in_use: false,
+            params: mldsa_params,
+            sigctx: sigctx,
+            data: None,
+        })
```

Comment:
these are completely same except for the line pubkey/privkey. Would it make sense to wrap them in some helper function to avoid code duplication?

Reply:
yeah, now they are, I will look at mergnign them later, if it is still the hase with the Hash variants